http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java

```diff
@@ -444,6 +461,22 @@ protected AuthChallenge resolveCode(String code) {
             TokenValidator.VerifiedTokens verifiedTokens = tokenValidator.parseAndVerifyToken(idTokenString, tokenString);
             idToken = verifiedTokens.getIdToken();
             token = verifiedTokens.getAccessToken();
+
+            String stateCookieValue = getCookieValue(deployment.getStateCookieName());
+            String[] cookieArr = stateCookieValue.split(";");
+            String sessionIdValue = "";
+            for(int i=0; i < cookieArr.length; i++) {
+                String cookie = cookieArr[i].trim();
+                if (cookie.startsWith(SESSION_ID+"=")) {
+                    sessionIdValue = cookie.substring((SESSION_ID+"=").length());
+                    break;
+                }
+            }
+            String nonceValue = idToken.getNonce();
+            if (nonceValue == null || !nonceValue.equals(String.valueOf(sessionIdValue.hashCode()))) {
```

Comment:
Here, we should use the cryptographic hash as mentioned above.

Reply:
Because the nonce validation is part of verifying the ID token, I think it makes sense for this check to be the last thing done in the OIDC `TokenValidator#parseAndVerifyToken` method. From the `TokenValidator` we have access to the `oidcClientConfiguration` from which we can get the value of the session random value cookie and we can get the nonce from `idJwtClaims`.

Reply:
done